sys/riscv/riscv/pmap.c

```diff
@@ -4204,7 +4275,7 @@ pmap_clear_modify(vm_page_t m)
 			va += VM_PAGE_TO_PHYS(m) - PTE_TO_PHYS(oldl2);
 			l3 = pmap_l2_to_l3(l2, va);
 			pmap_clear_bits(l3, PTE_D | PTE_W);
-			vm_page_dirty(m);
+			pmap_page_dirty(oldl2, m);
```

Comment:
This looks wrong

Reply:
I could readily believe, but can you articulate why?

We are demoting a superpage containing `m`, with PTE `oldl2`, and want to push `PTE_D` and `PTE_CD` up to `m`.  If you'd rather the `vm_page_dirty` not be conditional on `PTE_D` (it's not really conditional, in the sense that the `oldl2 & PTE_W` test implies `oldl2 & PTE_D`, as per the comment just above this block), I can inline `pmap_page_dirty` here.

Reply:
Hm, the use of `oldl2` looked suspect, but I see now why that's the case. However, I think the comment above the if needs updating to mention PTE_CD, and should we not have PTE_CD and PTE_CW in the call to `pmap_clear_bits`? In reality they don't matter to the architecture, but my understanding is you want at least the invariant `PTE_CW -> PTE_W`, if not others too.

Reply:
I'd have to think more if the same invariant (`PTE_W` implies `PTE_D`) holds true for `PTE_CW` and `PTE_CD`.  I don't really see anything in pmap_enter() or pmap_promote_l2() that explicitly guarantees the invariant for `PTE_W` and `PTE_D`.  Sure we would demote a writable-but-not-dirty superpage on the first fault, but I guess that pmap_clear_modify() wouldn't be called until the page was marked dirty and that is what prevents that.  That certainly won't be true for `PTE_CD` and `PTE_CW` here.

I think the right answer here is to actually leave `PTE_CD` and `PTE_CW` alone.  That is, I would just call `vm_page_dirty` here (so drop this hunk) and not try to bother with pushing `PTE_CD` into the VM page state.  pmap_clear_modify()'s goal isn't to clear CAP dirty state, it is to clear the non-CAP dirty state.